gpflow/features.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+import tensorflow as tf
+from functools import singledispatch
+from abc import ABCMeta, abstractmethod
+
+from .param import Param, Parameterized
+from . import conditionals
+from ._settings import settings
+
+
+float_type = settings.dtypes.float_type
+
+
+class InducingFeature(Parameterized, metaclass=ABCMeta):
+    """
+    Abstract base class for inducing features.
+    """
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """
+        Returns the number of features, relevant for example to determine the
+        size of the variational distribution.
+        """
+        pass
+
+    @abstractmethod
+    def Kuu(self, kern):
+        """
+        Calculates the covariance matrix between features for kernel `kern`.
+        """
+        pass
+
+    @abstractmethod
+    def Kuf(self, kern, Xnew):
+        """
+        Calculates the covariance matrix with function values at new points
+        `Xnew` for kernel `kern`.
+        """
+        pass
+
+
+class InducingPoints(InducingFeature):
+    """
+    Real-space inducing points
+    """
+
+    def __init__(self, Z):
+        """
+        :param Z: the initial positions of the inducing points, size M x D
+        """
+        super().__init__()
+        self.Z = Param(Z)
+        self._Mfeat = len(Z)
```

Comment:
Users are generally allowed to re-assign `Param` objects. This would render `self._mfeat` outdated. `__len__` needs to be derived either from `self.Z` directly, or updated whenever `self.Z` is re-assigned.

Reply:
Yes, I understand. This is a slightly difficult situation, but we have to support re-assigning `Param` objects. What do you think of the suggestion below, to maken `__len__` return a tensor when in tf_mode?

Reply:
Yeah, if that works reliably - agree that it's cleaner to not carry around potentially stale constants if it's easy to compute them on demand.

What do you mean by re-assigning Param objects? someone doing model.feat.Z = an array with more inducing points?

Reply:
Yes exactly. Although I think the correct syntax would be `model.feat.Z = Param(array_with_more_points)`.

It is quite useful actually. One may want to optimise inducing points, and then add more afterwards.

Reply:
What's the best way to check whether GPflow is in tf mode or not ?

Reply:
```
if self._tf_mode:
    stuff
```

@jameshensman?